Crash in TimerBase::stopSlowCase() via DNSResolveQueueCFNet::performDNSLookup

rniwa opened the pull request:
#### 59247a7628dd45966bd71f2a00523ed891650aa7
<pre>
Crash in TimerBase::stopSlowCase() via DNSResolveQueueCFNet::performDNSLookup
<a href="https://bugs.webkit.org/show_bug.cgi?id=291212">https://bugs.webkit.org/show_bug.cgi?id=291212</a>

Reviewed by Chris Dumez.

The crash appears to be caused by DNSResolveQueueCFNet::performDNSLookup updating WebCore timer
without grabbing WebThread lock. Speculatively fixed the crash by grabbing WebThread lock.

* Source/WebCore/platform/network/cf/DNSResolveQueueCFNet.cpp:
(WebCore::DNSResolveQueueCFNet::performDNSLookup):

Canonical link: <a href="https://commits.webkit.org/293419@main">https://commits.webkit.org/293419@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9ef5ffc7ce4a4f60ec8275c43e3e0c1b6108b374

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/98730 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/18363 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/8600 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/103855 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/49320 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/100775 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/18656 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/26814 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/75162 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/32312 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/101734 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/14172 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/89163 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/55519 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/13958 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/7128 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/48700 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/83907 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/7206 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/106226 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/25820 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/18828 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/84136 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/26197 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/85364 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/83623 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/28271 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/5949 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/19539 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/16071 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/25778 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/30960 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/25596 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/28916 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/27171 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

ddkilzer commented on Source/WebCore/platform/network/cf/DNSResolveQueueCFNet.cpp:
```diff
@@ -131,6 +131,10 @@ void DNSResolveQueueCFNet::performDNSLookup(const String& hostname, Ref<Completi
         if (status != nw_resolver_status_complete)
             return;
 
+#if PLATFORM(IOS_FAMILY)
+        WebThreadLock();
+#endif
```
I don't think this will fix the crash.  I think the timer needs to be scheduled on the WebThread, so this assertion on Line 123 is incorrect:

```
    RELEASE_ASSERT_WITH_MESSAGE(isMainThread(), "Always create timer on the main thread.");
```

Can we use something other than a `WebCore::Timer` here?

rniwa replied:
I don't think so. It's okay to schedule a WebCore::Timer in the main thread as long as WebThread lock is held.

rniwa committed:
Crash in TimerBase::stopSlowCase() via DNSResolveQueueCFNet::performDNSLookup
https://bugs.webkit.org/show_bug.cgi?id=291212

Reviewed by Chris Dumez.

The crash appears to be caused by DNSResolveQueueCFNet::performDNSLookup updating WebCore timer
without grabbing WebThread lock. Speculatively fixed the crash by grabbing WebThread lock.

* Source/WebCore/platform/network/cf/DNSResolveQueueCFNet.cpp:
(WebCore::DNSResolveQueueCFNet::performDNSLookup):

Canonical link: https://commits.webkit.org/293419@main
